### DISCUSSION TITLE
uploadBlob by CID

### DISCUSSION THREAD

**Discussion by DavidBuchanan314** [bluesky-social/atproto#1756] [Protocol (atproto)]
If GIF support happens, or even if it doesn't, there are going to be certain blobs that get reposted a lot.

It's already easy for a PDS to save on storage by deduplicating blobs - which comes for free if you have a CID-indexed blob store.

However, there's still a waste of bandwidth, every time someone re-uploads the same blob.

What if there was an API with the same effect as `uploadBlob`, but taking a CID as argument rather than the actual file data? If the PDS already has a blob with that CID, it acts as if it had been uploaded again, by the current user, but without having to send it over the wire. If the PDS doesn't recognize the CID, it returns an error response, and the client can upload it normally.

Pros:

- Bandwidth savings.

Cons:

- Extra network round-trip in the case that the PDS doesn't already have the CID (which is probably the most common outcome).

- Extra implementation effort.

- Currently, you need a (DID, CID) tuple to check if a PDS hosts a given blob, and/or to retrieve it (via `getBlob`). This proposed feature would allow probing via CID only. Perhaps this is undesirable?

**Comment by bnewbold** [comment]
Where we are roughly at now (as you noted) is having moved from PDS-global blobs (by CID) to repo-local. The main motivation was to make life-cycle management easier: deletions of referencing records, or deletions of entire accounts. The actual blob backend storage *could* be de-duped with a secondary system/index, but also so far there has been very little actual de-dupe, even with the broad use of meme images. There are also some corner-case situations where a blob might be unacceptable (takedown) in some post contexts, but not others, so tracking per-repo mod state (like hard blob takedown) could be desirable.

A related idea would be things like "sticker packs", a la messaging apps, which could be done with pre-announced CIDs, or some other form of indirection. I guess blobs can't currently be references cross-repo, but maybe AT URI syntax could be extended to enable that.

Website social cards might get an efficiency boost: linking and getting the same generic placeholder image over and over is wasteful. But the client already had to download and hash anyways.

My memory is that Dropbox did cross-account de-dupe in the early days, to skip unnecessary uploads, but ripped it out over privacy and abuse issues. Those mostly don't exist for current atproto uses, and we'll probably need to do something different anyways for private/group-private data.
